feat: DPROD-687 ability to sync jira boards

kpiwko opened the pull request:
* now can also sync jira boards
* additional tests for settings PageObject

kpiwko commented on logilica_weekly_report/page_settings.py:
```diff
@@ -20,29 +22,39 @@ class SettingsPage:
 
-    def open_integration_configuration(self, integration: str, connector: str) -> None:
+    def open_integration_configuration(
+        self, *, integration: str, connector: str
```
sure, I'm basically using named parameters for function with more than two arguments, so I only aligned it with the rest of the file.

webbnh commented on tests/test_page_settings.py:
```diff
@@ -0,0 +1,102 @@
+
+
+def has_entity_id_imported_side_effect(*args, **kwargs):
+    if "miss" in kwargs.get("text", "") and kwargs.get("exact") is True:
+        mock_for_missing_entry = MagicMock()
+        # chain: .nth(0).is_visible() => False
+        mock_for_missing_entry.nth.return_value.is_visible.return_value = False
+        return mock_for_missing_entry
+    else:
+        mock_for_existing_entry = MagicMock()
+        # chain: .nth(0).is_visible() => True
+        mock_for_existing_entry.nth.return_value.is_visible.return_value = True
+        return mock_for_existing_entry
```
Alternatively,
```suggestion
    result = "miss" in kwargs.get("text", "") and kwargs.get("exact", False) is True
    mock_for_missing_entry = MagicMock()
    # chain: .nth(0).is_visible() => result
    mock_for_missing_entry.nth.return_value.is_visible.return_value = result
    return mock_for_missing_entry
```

kpiwko replied:
Almost, there needs to be `not result` ;-)

webbnh commented on tests/test_page_settings.py:
```diff
@@ -0,0 +1,102 @@
+
+        mock_locator = MagicMock()
+        mock_locator.count.return_value = 1
+        page_mock.get_by_role.return_value.filter.return_value.get_by_role.return_value = (
+            mock_locator
+        )
```
I would use a temporary variable to avoid Black adding parentheses:
```suggestion
        frv = page_mock.get_by_role.return_value.filter.return_value
        frv.get_by_role.return_value = mock_locator
```
Ditto at line 90.

webbnh commented on tests/test_page_settings.py:
```diff
@@ -0,0 +1,102 @@
+        mock_expect.return_value = page_mock
+        page = SettingsPage(page=page_mock)
+        page.sync_integrations(integrations=JIRA_INTEGRATION)
+        print(mock_locator.mock_calls)
+        page_mock.get_by_role.assert_any_call("heading", name="Jira Settings")
```
Should that `print()` call be removed now that things are working?

kpiwko replied:
indeed

webbnh commented on tests/test_page_settings.py:
```diff
@@ -0,0 +1,102 @@
+        with raises(RuntimeError) as error:
+            page.sync_integrations(integrations=MISSING_ENTITIES)
+
+        assert (
+            sum(len(value_list) for value_list in error.value.args[0].values()) == 2
+        ), "There are two integrations with 2 failures that failed to sync"
```
We only see the supplied text when the assertion _fails_, which would be when the number of integration failures is **_not_** two, right?  So, I think the supplied text should probably say _that_.  Ideally, the assertion should report what the expected value was and what the actual value was.

Also, should we be making an assertion about the user-visible text of the `RuntimeError` (i.e., to confirm that the `RuntimeError` instance that we received was the expected one)?

kpiwko replied:
We don't need to do that, if there are sync failures, it raises `RuntimeError(sync_failures)`

So `error.value.args[0]` returns `defaultdict[Tuple[str, str], list[Tuple[str, str]]]` and that's where we get `.values()`. Other RuntimeError instance would not have that.

webbnh commented on logilica_weekly_report/configuration_schema.py:
```diff
@@ -42,6 +42,10 @@
                         "type": "array",
                         "items": {"type": "string"},
                     },
+                    "membership_boards": {
+                        "type": "array",
+                        "items": {"type": "string"},
+                    },
```
Should any of the test scenarios in [`test_configuration_schema.py`](https://github.com/developerproductivity/logilica-weekly-report/blob/main/tests/test_configuration_schema.py) be tweaked for this change?

kpiwko replied:
I have updated that now, :-)

webbnh reviewed (approved):
Looks good.  However, I have a few nits, some suggestions, and a couple of pointed questions.

kpiwko commented on logilica_weekly_report/page_settings.py:
```diff
@@ -20,29 +22,39 @@ class SettingsPage:
 
-    IMPORTED_REPOSITORY_TIMEOUT = 9000
-    AVAILABLE_REPOSITORY_TIMEOUT = IMPORTED_REPOSITORY_TIMEOUT * 6
+    IMPORT_TIMEOUT = 9000
+    AVAILABLE_LIST_TIMEOUT = IMPORT_TIMEOUT * 6
```
You are right, there is no dependency here, I'll make this independent..

kpiwko commented on logilica_weekly_report/page_settings.py:
```diff
@@ -161,19 +105,34 @@ def sync_integrations(self, integrations: dict[str, Any]) -> None:
             )
-            # process membership repositories
-            self.process_membership_repositories(
+            # process membership based repositories
+            self.process_repositories(
```
There were actually 4 occurrences :-)

kpiwko commented on logilica_weekly_report/page_settings.py:
```diff
@@ -188,78 +147,216 @@ def sync_integrations(self, integrations: dict[str, Any]) -> None:
+        check_function: Callable[[list[str]], list[str]],
+        failures: IntegrationSyncFailures,
+    ) -> None:
+        """Makes sure that entities (could be a repository slugs, board names)
+        are configured for given integration.
```
there should have been singulars instead of plurals, I have fixed it that way.

kpiwko commented on logilica_weekly_report/page_settings.py:
```diff
@@ -188,78 +147,216 @@ def sync_integrations(self, integrations: dict[str, Any]) -> None:
-            for added_repo in repositories
-            if not self.has_repository_imported(added_repo)
+        self.page.wait_for_timeout(self.IMPORT_TIMEOUT)
+        missing_entries = [
+            (entry, f"❌{entity_type} {entry} was not imported.")
+            for entry in entity_ids
+            if not search_function(entity_id=entry)
         ]
```
I like this however I find it problematic if a function has multiple parameters of the same type (e.g. `str`). This is the reason why I have introduced keyword based args.

kpiwko commented:
@webbnh please check, I have addressed your comments with the exception of `filter` and `filterfalse`, as I don't want to change the API to have 2 `str` args only distinguished by their position and using lambda does not increase readability. This has hit me multiple times in the past.

kpiwko commented:
@webbnh addressed your feedback in #20 and #21, please have a look. Thanks!

webbnh reviewed (dismissed):
It looks good.

I have a couple of stylistic comments on your updates, and it looks like you didn't act on [one of my comments from #21](https://github.com/developerproductivity/logilica-weekly-report/pull/21#discussion_r1999793574).  Let me know if you're not interested, and I'll resolve the conversations.

kpiwko commented:
@webbnh addressed remaining feedback, thanks for the suggestions.

webbnh reviewed (approved):
Ship it!
